src/API.jl

```diff
@@ -114,13 +114,15 @@ end
 
 function check_package_name(x::AbstractString, mode=nothing)
```

Comment:
I'm still a bit confused about whether this PR does everything you want it to. If not, one "cheap" way to get better inference is to make the signature `check_package_name(x::String, mode::Union{Nothing,SomeType}=nothing)` and then also provide an argument-standardizing method
```julia
check_package_name(x::AbstractString, mode=nothing) = check_package_name(String(x)::String, convert(Union{Nothing,SomeType}, mode)::Union{Nothing,SomeType})
```
The typeasserts guarantee you'll never get a StackOverflowError.